test/core/xds/xds_client_test.cc

```diff
@@ -3991,48 +4033,125 @@ TEST_F(XdsClientTest, FallbackReportsError) {
   ASSERT_NE(resource, nullptr);
   EXPECT_EQ(resource->name, "foo1");
   EXPECT_EQ(resource->value, 6);
+  request = WaitForRequest(stream.get());
+  ASSERT_TRUE(request.has_value());
+  CheckRequest(*request, XdsFooResourceType::Get()->type_url(),
+               /*version_info=*/"1", /*response_nonce=*/"A",
+               /*error_detail=*/absl::OkStatus(),
+               /*resource_names=*/{"foo1"});
   TriggerConnectionFailure(primary_server,
                            absl::UnavailableError("Server down"));
+  stream->MaybeSendStatusToClient(absl::UnavailableError("Server is down"));
+  stream = WaitForAdsStream();
+  ASSERT_NE(stream, nullptr);
   // XdsClient should have sent a subscription request on the ADS stream.
   request = WaitForRequest(stream.get());
   ASSERT_TRUE(request.has_value());
   CheckRequest(*request, XdsFooResourceType::Get()->type_url(),
-               /*version_info=*/"1", /*response_nonce=*/"A",
+               /*version_info=*/"1", /*response_nonce=*/"",
                /*error_detail=*/absl::OkStatus(),
                /*resource_names=*/{"foo1"});
   auto error = watcher->WaitForNextError();
   ASSERT_TRUE(error.has_value());
   EXPECT_EQ(error->message(),
             "xDS channel for server default_xds_server: Server down (node "
             "ID:xds_client_test)");
+  TriggerConnectionFailure(fallback_server,
+                           absl::UnavailableError("Another server down"));
+  stream->MaybeSendStatusToClient(
+      absl::UnavailableError("Another server down"));
+
   // Fallback happens now
-  auto watcher2 = StartFooWatch("foo2");
-  // Requested on stream1
-  request = WaitForRequest(stream.get());
+  auto watcher2 = StartFooWatch("foo1");
+  // XdsClient should have sent a subscription request on the ADS stream.
+  error = watcher->WaitForNextError();
+  ASSERT_TRUE(error.has_value());
+  EXPECT_THAT(error->message(), ::testing::ContainsRegex("Another server down"))
+      << error->message();
+  resource = watcher2->WaitForNextResource();
+  ASSERT_NE(resource, nullptr);
+  EXPECT_EQ(resource->name, "foo1");
+  EXPECT_EQ(resource->value, 6);
+  error = watcher2->WaitForNextError();
+  ASSERT_TRUE(error.has_value());
+  EXPECT_THAT(error->message(),
+              ::testing::ContainsRegex("Another server down"));
+  CancelFooWatch(watcher.get(), "foo1");
+  CancelFooWatch(watcher2.get(), "foo1");
+}
+
+TEST_F(XdsClientTest, FallbackOnStartup) {
+  FakeXdsBootstrap::FakeXdsServer primary_server;
+  FakeXdsBootstrap::FakeXdsServer fallback_server("fallback_xds_server");
+  // Regular operation
+  InitXdsClient(FakeXdsBootstrap::Builder().SetServers(
+      {primary_server, fallback_server}));
+  // Start a watch for "foo1".
+  auto watcher = StartFooWatch("foo1");
```

Comment:
This input will cause the client to send a request on the stream before it sees the connectivity failure.